docs/source/coding_guidelines.md

```diff
@@ -0,0 +1,213 @@
+
+### Static Analysis and Linting
+
+We use `flake8` as a linter, and `mypy` and pyright for static typechecking. Both tools run as part of the PR build, and must run without errors for a contribution to be accepted. `mypy` requires that all functions and methods carry type annotations, see [mypy documentation](https://mypy.readthedocs.io/en/latest/getting_started.html#function-signatures-and-dynamic-vs-static-typing)
+
+We highly recommend to run all those tools _before_ pushing the latest changes to a PR. If you have `make` installed, you can run both tools in one go via `make check` (from the repository root folder)
+
+### String Style
+
+End sentences in docstrings with a period:
+
+```python
+def method(self, arg: int) -> None:
+    """
+    Method description, followed by an period.
+    """
+```
+
+* End strings in error messages with a period.
+
+### Documentation
+
+We are using Sphinx for generating documentation. We recommend using a VSCode extension for auto-generating documentation templates, for example `njpwerner.autodocstring` (this extension is already pre-configured in VSCode's workspace settings file).
+
+Common mistakes when writing docstrings:
+
+* There must be a separating line between a function description and the documentation for its parameters.
+* In multi-line parameter descriptions, continuations on the next line must be indented.
+* Sphinx will merge the class description and the arguments of the constructor `__init__`. Hence, there is no need to
+  write any text in the constructor docstring other than the constructor arguments.
+* Use `>>>` to include code snippets. PyCharm will run intellisense on those to make authoring easier.
+* To generate the Sphinx documentation on your dev machine, run `make html` in the `./docs` folder, and then
+  open `./docs/build/html/index.html`
+
+Example:
+
+```python
+class Foo:
+    """
+    This is the class description.
+
+    The following block will be pretty-printed by Sphinx. Note the space between >>> and the code!
+    
+    Usage example:
+        >>> from module import Foo
+        >>> foo = Foo(bar=1.23)
+    """
+
+    ANY_ATTRIBUTE = "what_ever."
+    """Document class attributes after the attribute."""
+
+    def __init__(self, bar: float = 0.5) -> None:
+        """
+        :param bar: This is a description for the constructor argument.
+            Long descriptions should be indented.
+        """
+        self.bar = bar
+
+    def method(self, arg: int) -> None:
+        """
+        Method description, followed by an empty line. Argument names like ``arg`` are rendered nicely
+        if enclosed in double backtick.
+
+        This method can raise a :exc:`ValueError`.
+        
+        :param arg: This is a description for the method argument.
+            Long descriptions should be indented.
+        """
+```
+
```

Comment:
This is a little primer, but we should add a link to a Sphinx quickstart or the full reference, I think.